src/main/java/com/microsoft/aad/adal4j/AuthenticationContext.java

```diff
@@ -282,6 +283,92 @@ public AuthenticationResult call() throws Exception {
+            final AuthenticationCallback callback) {
+
+        return service.submit(new Callable<AuthenticationResult>() {
+
+            private String userName;
+            private String resource;
+            private ClientAuthentication clientAuth;
+            private ClientDataHttpHeaders headers;
+
+            @Override
+            public AuthenticationResult call() throws Exception {
+                AuthenticationResult result = null;
+                try {
+                    AdalAuthorizatonGrant authGrant = new AdalAuthorizatonGrant(getAuthorizationGrantIntegrated(this.userName), this.resource);
+                    result = acquireTokenCommon(authGrant, this.clientAuth, this.headers);
+                    logResult(result, headers);
+                    if (callback != null) {
+                        callback.onSuccess(result);
+                    }
+                }
+                catch (final Exception ex) {
+                    log.error(LogHelper.createMessage("Request to acquire token failed.", this.headers.getHeaderCorrelationIdValue()), ex);
+                    if (callback != null) {
+                        callback.onFailure(ex);
+                    }
+                    else {
+                        throw ex;
+                    }
+                }
+                return result;
+            }
+
+            private Callable<AuthenticationResult> init(String userName,
+                    final String resource,
+                    final ClientAuthentication clientAuth,
+                    final ClientDataHttpHeaders headers) {
+                this.userName = userName;
+                this.resource = resource;
+                this.clientAuth = clientAuth;
+                this.headers = headers;
+                return this;
+            }
+
+        }.init(userName, resource, clientAuth, new ClientDataHttpHeaders(this.getCorrelationId())));
+    }
+
+    AuthorizationGrant getAuthorizationGrantIntegrated(String userName) throws Exception {
+        AuthorizationGrant updatedGrant = null;
+
+        String userRealmEndpoint = authenticationAuthority.getUserRealmEndpoint(URLEncoder.encode(userName, "UTF-8"));
+
+        // Get the realm information
+        UserDiscoveryResponse userRealmResponse = UserDiscoveryRequest.execute(userRealmEndpoint, proxy, sslSocketFactory);
+
+        if (userRealmResponse.isAccountFederated() && "WSTrust".equalsIgnoreCase(userRealmResponse.getFederationProtocol())) {
+            String mexURL = userRealmResponse.getFederationMetadataUrl();
+            String cloudAudienceUrn = userRealmResponse.getCloudAudienceUrn();
+
+            // Discover the policy for authentication using the Metadata Exchange Url.
+            // Get the WSTrust Token (Web Service Trust Token)
+            WSTrustResponse wsTrustResponse = WSTrustRequest.execute(mexURL, cloudAudienceUrn, proxy, sslSocketFactory);
+
+            // Make the OAuth2 call to get the access Token.

```

Comment:
this comment is quite misleading - there is no OAuth2 call here

Reply:
thank you. Just moved the comment to right place